probe-rs/src/bin/probe-rs/cmd/dap_server/server/core_data.rs

```diff
@@ -383,4 +383,159 @@ impl<'p> CoreHandle<'p> {
         }
         Ok(())
     }
+
+    /// Traverse all the variables in the available stack frames, and return the memory ranges
+    /// required to resolve the values of these variables. This is used to provide the minimal
+    /// memory ranges required to create a [`CoreDump`] for the current scope.
+    pub(crate) fn get_memory_ranges(&mut self) -> Vec<Range<u64>> {
+        let mut all_discrete_memory_ranges = Vec::new();
+        // Expand and validate the static and local variables for each stack frame.
+        for frame in self.core_data.stack_frames.iter_mut() {
+            let mut variable_caches = Vec::new();
+            if let Some(static_variables) = &mut frame.static_variables {
+                variable_caches.push(static_variables);
+            }
+            if let Some(local_variables) = &mut frame.local_variables {
+                variable_caches.push(local_variables);
+            }
+            for variable_cache in variable_caches {
+                // Cache the deferred top level children of the of the cache.
+                variable_cache.recurse_deferred_variables(
+                    &self.core_data.debug_info,
+                    &mut self.core,
+                    None,
+                    &frame.registers,
+                    frame.frame_base,
+                    10,
+                    0,
+                );
+                all_discrete_memory_ranges.append(&mut variable_cache.get_discrete_memory_ranges());
+            }
+            // Also capture memory addresses for essential registers.
+            for register in frame.registers.0.iter() {
+                if let Ok(Some(memory_range)) = register.memory_range() {
+                    all_discrete_memory_ranges.push(memory_range);
+                }
+            }
+        }
+        // Consolidating all memory ranges that are withing 0x400 bytes of each other.
+        consolidate_memory_ranges(&mut all_discrete_memory_ranges, 0x400)
+    }
+}
+
+/// Return a Vec of memory ranges that consolidate the adjacent memory ranges of the input ranges.
+/// Note: The concept of "adjacent" is calculated to include a gap of up to specicied number of bytes between ranges.
+/// This serves to consolidate memory ranges that are separated by a small gap, but are still close enough for the purpose of the caller.
+pub(crate) fn consolidate_memory_ranges(
+    discrete_memory_ranges: &mut Vec<Range<u64>>,
+    include_bytes_between_ranges: u64,
+) -> Vec<Range<u64>> {
```

Comment:
Why does this take a mutable reference to a `Vec`, but returns a `Vec` as well? If `discrete_memory_ranges` gets changed, it should probably just be consumed.

Reply:
@Tiwalun ... I'm stuck here. Can you help with a suggestion please? 

I've changed it to the following, but then the compiler complains about trying to do a mutable borrow twice (when I drain, and when I push). How would you solve this?

```
fn consolidate_memory_ranges(
    memory_ranges: &mut Vec<Range<u64>>,
    include_bytes_between_ranges: u64,
) {
    memory_ranges.sort_by_cached_key(|range| (range.start, range.end));
    memory_ranges.dedup();
    let mut condensed_range: Option<Range<u64>> = None;

    for memory_range in memory_ranges.drain(..) {
        if let Some(range_comparitor) = condensed_range {
            if memory_range.start <= range_comparitor.end + include_bytes_between_ranges + 1 {
                let new_end = std::cmp::max(range_comparitor.end, memory_range.end);
                condensed_range = Some(Range {
                    start: range_comparitor.start,
                    end: new_end,
                });
            } else {
                memory_ranges.push(range_comparitor);
                condensed_range = Some(memory_range.clone());
            }
        } else {
            condensed_range = Some(memory_range.clone());
        }
    }

    if let Some(range_comparitor) = condensed_range {
        memory_ranges.push(range_comparitor);
    }
}
```